### DISCUSSION TITLE
サボタージュの起動と解消のSystemTypeとamountについて

### DISCUSSION THREAD

**Discussion by yurinakira** [yukieiji/ExtremeRoles#176] [Q&A]
適切な投稿先(Discord)が見つからなかったのでこちらに失礼します。
不適切であれば確認後クローズをお願いします。

私はTOHの開発者メンバーのゆりのと申します。
Twitterで
「多分だけどサボタージュの発動と修理もぜんっぶRepairSystemってやつに一任されてるな？」と呟かれているのを見たのでこちらに情報連携をしようと思いissueを切らせていただきました。

■まずはじめに
個人的な趣味で「サボの開始」と「サボの解消」を検知できないかとホストmodであるTOHで調べたものなので、
クライアントmodでは結果が異なることがあるかもしれません。

後述のsystemTypeとamountに関しては以下の引数の事を指します。
```
public static bool Prefix(ShipStatus __instance,
            [HarmonyArgument(0)] SystemTypes systemType,
            [HarmonyArgument(1)] PlayerControl player,
            [HarmonyArgument(2)] byte amount)
```

「サボの発動」「サボの解消」の検知に関しては、
実装はしてみましたが、現状動作確認がまだなのと、
コードが力業なのと、完全に趣味でやっていて凄く出来が悪く
滅茶苦茶恥ずかしいので載せていません...。

お察しの通りサボの発動と解消は同じハーモニーパッチです。↓
[HarmonyPatch(typeof(ShipStatus), nameof(ShipStatus.RepairSystem))]

■systemTypeとamountの組み合わせについて
Skeld
![image](https://user-images.githubusercontent.com/84663668/187565640-de71baf2-770d-4eaa-abf8-4956ecb0bc96.png)
MIRA HQ
![image](https://user-images.githubusercontent.com/84663668/187565672-0b5a514f-770c-4227-b58d-440b7369c63c.png)
Polus
![image](https://user-images.githubusercontent.com/84663668/187565720-7c42ae77-2ee1-4b55-a249-d47c5e1996d6.png)
AirShip
![image](https://user-images.githubusercontent.com/84663668/187565742-9b28900c-e517-44d4-9f09-1d42a44f2cd6.png)

■停電の解消について
SystemTypes がElectrical amountが0~4の時に、
以下がfalseだったら停電解消と判断するようにすればいけると思います。(動作未確認)
ShipStatus.Instance.Systems[SystemTypes.Electrical].Cast<SwitchSystem>().IsActive

**Comment by yurinakira** [comment]
ShipStatus.Instance.Systems[SystemTypes.Electrical].Cast().IsActiveの値の監視をすればもっと簡単に対応可能かも知れませんので、何かいい方法があったら教えていただけると助かります。。

*上記の値監視だと「誰がサボ解消したか」の情報が取れないので、私は今回の調査のようにRepairSystemで判定入れてみました。

**Comment by yukieiji** [comment]
とりあえずIssuesで話す内容ではないのでDiscussionのQ＆Aに移動します
タグは後で整備します

**Reply by yurinakira** [reply]
Discussionへの移動ありがとうございます。(Discussion初めて触るので不手際あったら指摘お願いします。)

**Comment by yukieiji** [comment]
とりあえず上から

> 適切な投稿先(Discord)が見つからなかったのでこちらに失礼します。

Discordのリンクは一応このリポジトリのReadmeにもあります
こっちで立てる場合はDiscussionでしょうか

> 個人的な趣味で「サボの開始」と「サボの解消」を検知できないかとホストmodであるTOHで調べたものなので、
クライアントmodでは結果が異なることがあるかもしれません。

サボの解消については[メンテナーのコード](https://github.com/yukieiji/ExtremeRoles/blob/master/ExtremeRoles/Roles/Solo/Crewmate/Maintainer.cs#L45-L105)でやってるRpcRepairDamageでやってるのでそこの内部で呼んでるRepairSystemにホックして値を監視すれば取れると思ってます

停電に関してはスイッチの上げ下げがamountに確か格納されていてそれが全部一定値にならないと修理したという判定にならないのでMOD側で独自RPCを飛ばしてSwitchSystemの値を強制的に書き換えてます

> ShipStatus.Instance.Systems[SystemTypes.Electrical].Cast().IsActiveの値の監視をすればもっと簡単に対応可能かも知れませんので、何かいい方法があったら教えていただけると助かります。。
*上記の値監視だと「誰がサボ解消したか」の情報が取れないので、私は今回の調査のようにRepairSystemで判定入れてみました。

サボ周りはRepairSystemとPlayerTaskを監視するのが一番楽だと思います(RepairSystemでプレイヤーを取って、PlayerTaskで発動、解消と種類をとる)

IsActiveはたしか別のサボタージュを発動させないためのトリガーだったりするのでドアとかだと使われてなかったはずです

PS.RepairSystemでした間違えました

**Reply by yurinakira** [reply]
Discordありがとうございます。見つけましたので参加させていただきました。
READMEで「Discord」で検索かけたら引っかからなかったのでないものかと思っていました。。。


RepairDamage見れてないですが、たぶんいけると思います。


停電に関しては、
ShipStatus.RepairSystemでは、配電盤左から0~4(触った場所)でamountの値が帰って来る挙動になっていました。
RepairDamageでは未確認です。
今度確認してみようと思います。
